### PR TITLE
spython.main.Client.build: add sudo_options option

### PR DESCRIPTION
**Pull request by tschoonj**
This allows me to pass '--preserve-env=SINGULARITY_CACHEDIR,SINGULARITY_TMPDIR' to sudo,
which I need to deal with my very small /root and /tmp partitions.

### CODE REVIEW

File: spython/utils/terminal.py [hunk 1]
```diff
@@ -103,9 +103,14 @@ def stream_command(cmd, no_newline_regexp="Progess", sudo=False):
        cmd: the command to send, should be a list for subprocess
        no_newline_regexp: the regular expression to determine skipping a
                           newline. Defaults to finding Progress
+       sudo_options: string or list of strings that will be passed as options to sudo
 
     '''
-    if sudo:
+    if sudo and sudo_options is not None:
+        if isinstance(sudo_options, str):
+            sudo_options = sudo_options.split()
```
Comment (vsoch) [inline spython/utils/terminal.py]: For this, I think we should use shlex to parse the string. See shlex.split https://docs.python.org/3/library/shlex.html

File: spython/utils/terminal.py [hunk 2]
```diff
@@ -140,9 +146,14 @@ def run_command(cmd,
        capture: if True, don't set stdout and have it go to console. This
                 option can print a progress bar, but won't return the lines
                 as output.
+       sudo_options: string or list of strings that will be passed as options to sudo
     '''
 
-    if sudo:
+    if sudo and sudo_options is not None:
```
Comment (vsoch) [inline spython/utils/terminal.py]: I'm seeing duplicate code :)

Comment (vsoch) [reply spython/utils/terminal.py]: In the same way that we have self._init_command() perhaps there could be a function to handle sudo options, that way if there is a bug fix or similar, we don't have to do it in two places. And do you see sudo options being useful in other functions, aside from build?

File: spython/utils/terminal.py (hunk 1 above)
Comment (tschoonj) [reply spython/utils/terminal.py]: Done

File: spython/utils/terminal.py (hunk 2 above)
Comment (tschoonj) [reply spython/utils/terminal.py]: Duplicate code replaced with a function

Comment (tschoonj) [reply spython/utils/terminal.py]: I am not sure about sudo usage elsewhere, I had never used `spython` before this afternoon so not familiar with it really :smile:

Comment (tschoonj) [reply spython/utils/terminal.py]: I assume I need to bump the version and update the changelog?

Anything else?

Comment (vsoch) [reply spython/utils/terminal.py]: Welcome to the library! :D Glad to have you here too :)

**Comment by vsoch** [comment]
I think we are forgetting one important thing... *looks at CHANGELOG.md and version.py...*

**Comment by tschoonj** [comment]
Pushed that commit just now...

**Comment by vsoch** [comment]
Nope that is it! I think this is definitely needed, and it's good to handle the positioning and logic of adding the arguments with the base functions. Otherwise, we'd have to do the same checks (is the user using sudo? Are there sudo args?) in all the calling functions, and it would be redundant. So - it's a choice between having the lowest level functions (stream and run command) to have specialized options, *or* having redundant code. I like the first approach better, and based on your PR, I see that you do too.

**Comment by tschoonj** [comment]
Yeah, I kind of went down the rabbit hole until I found where the Popen magic happens :rabbit:

**Comment by tschoonj** [comment]
By the way, your Circle-CI Orb setup is working really nicely for me... Thanks for that!

I will now be using `spython` to build the Singularity images and use `sregistry` to push them to an S3 bucket, all using pure Python... 

One thing I was wondering about... Do you think it would be a good idea to add the singularity installation to the cache as well? Now it is getting rebuilt every single time. Not hugely time consuming but still...

**Comment by vsoch** [comment]
That would be a great idea! I hadn't tackled it yet because I was worried about how to install the different versions alongside one another while still using a cache (for example, I've had Python build setups where I was caching Anaconda, and I *thought* I was testing Python 2 and 3, but in fact was just testing one, and the solution was to move them to different paths in the same cache. I believe there is an issue open for this #126 - do you want to either take a shot, or chat about how we could go about it (and then I'll try?) We would want to discuss if the cache should happen here, or with the Orb, and then how to properly do it. Anyway this PR is good to go - merging here and let's pick up conversation in #126!

**Comment by vsoch** [comment]
https://pypi.org/project/spython/0.0.73/

**Pull request merged**